script_facade/models/r4/medication_request.py

```diff
@@ -16,51 +16,77 @@ def __init__(self):
         self.source_identifier = None
 
     @classmethod
-    def from_xml(cls, med_dispensed, source_identifier):
+    def from_xml(cls, med_dispensed, source_identifier, script_version):
         """Build a MedicationRequest from a MedicationDispensed xml element object
 
         :param med_dispensed: MedicationDispensed xml element object
 
         """
-        med_order = cls()
-        med_order.source_identifier = source_identifier
 
-        # todo: separate finding/extract into separate steps
-        drug_description = med_dispensed.xpath('.//DrugDescription/text()')[0]
+        # wrap version-specific extraction methods
+        if script_version == "106":
+            med_request = cls.from_106_xml(med_dispensed)
+        elif script_version == "20170701":
+            med_request = cls.from_20170701_xml(med_dispensed)
+        else:
+            # TODO raise exception
+            pass
 
+        med_request.source_identifier = source_identifier
+        med_request.authored_on = cls.authored_on_from_xml(med_dispensed)
 
-        drug_coded = med_dispensed.xpath('.//DrugCoded')[0]
-        product_code = drug_coded.xpath('.//ProductCode/text()')[0]
-        product_code_qualifier = drug_coded.xpath('.//ProductCodeQualifier/text()')[0]
 
+        return med_request
 
-        # attempt code system lookup
-        product_code_qualifier = drug_code_system_map.get(product_code_qualifier, product_code_qualifier)
 
-        # element no longer present in OHP PDMP response
-        #strength = drug_coded.xpath('.//*[local-name()="Strength"]')[0].text
+    @classmethod
+    def from_20170701_xml(cls, med_dispensed):
+        med_request = cls()
+        med_request.medication = cls.med_cc_from_20170701_xml(med_dispensed)
+        med_request.requester = cls.requester_from_20170701_xml(med_dispensed)
+        med_request.dispense_request = cls.dispense_request_from_20170701_xml(med_dispensed)
 
+        return med_request
 
-        med_order.authored_on = med_dispensed.xpath('.//WrittenDate/Date/text()')[0]
-        med_cc = {
-            'coding': [{
-                'system': product_code_qualifier,
-                'code': product_code,
-                'display': drug_description,
-            }],
-            'text': drug_description,
-        }
 
-        med_order.medication = med_cc
+    @classmethod
+    def from_106_xml(cls, med_dispensed):
```

Comment:
You've done a great job of isolating the method differences between the two versions.
It'd be cleaner to use inheritance here, such that `class MedicationRequest106(MedicationRequest)`  only defines `for_xml()` and other methods currently in `MedicationRequest` containing version numbers.